libr/reg/reg.c

```diff
@@ -200,7 +200,7 @@ R_API void r_reg_free_internal(RReg *reg, bool init) {
 static int regcmp(RRegItem *a, RRegItem *b) {
 	int offa = (a->offset * 16) + a->size;
 	int offb = (b->offset * 16) + b->size;
-	return offa > offb;
+	return (offa > offb) - (offa < offb);
```

Comment:
I believe the problem of the registers in the wrong order is because of this.
Try this:
```diff
-       int offa = (a->offset * 16) + a->size;
-       int offb = (b->offset * 16) + b->size;
-       return (offa > offb) - (offa < offb);
+       int off_diff = (a->offset < b->offset) - (a->offset > b->offset);
+       if (off_diff) {
+               return off_diff;
+       }
+       return (a->size > b->size) - (a->size < b->size);
```

Reply:
It's just sorted differently (registers in reverse order by offset, smallest size first), I'm not convinced it's a correct solution. Look at how the zignatures change for example. And for architectures that have less than 43 registers, it will be reversed. IMHO sorting the anal vars by name on print might be better. Don't forget the previous order was accidental.

Another possiblity: assign "delta" of register anal vars to be specific to the current call convention, and not derivative of the register offset.

Reply:
Yes it just stores the registers in reverse order, but it seems to me that's more or less what was intended before (though ofc the ordering was wrong at the end).

> And for architectures that have less than 43 registers, it will be reversed.

Why? It should behave the same for all number of registers.

> IMHO sorting the anal vars by name on print might be better

No, not by name because you want to see what is the actual order of the vars/argument. You want to know that the first argument you see is actualy the first argument passed to the function.

> assign "delta" of register anal vars to be specific to the current call convention, and not derivative of the register offset.

I think this is the proper solution long term, yes.

> I'm not convinced it's a correct solution. Look at how the zignatures change for example. 

I will have a look, but I'm not convinced of current PR's output either :(

Reply:
If this fix is appropriate for the x86_64 call conventions, that's fine for me. My initial motivation for this PR was to keep the zignatures stable for arm32 after PR #17462 ; I then discovered the discrepancy between the two sorts, and attempted to fix it. It would have been shorter to just fix the zignatures in the previous PR. I don't have a preference as I don't rely on this feature.

Note: reversing the register order, will just change the zignatures for all architectures, but that's fine if it's a one-time thing, I guess.

Reply:
Also, for architectures that pass arguments in an other order (e.g arm32/aarch64), the argument order will be reversed.

Reply:
> Also, for architectures that pass arguments in an other order (e.g arm32/aarch64), the argument order will be reversed.

good point!

I did not intend to slow you down, but what sometimes seems like a simple fix then risk breaking other stuff, that's why I was worried about the tests changes (and way too often we accidentally fix tests in the wrong way :( )

Anyway, it seems there is no simple fix here, so I'm ok as it is. Thanks!

Reply:
Please note that I really appreciated your reviews. Maintaining constant high quality in an open-source project is very hard, and I appreciate you taking the time to look at this PR.